### PR TITLE
[Bounties] Critical Module Ravamp

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by brebenelmihnea** [comment]
I'll add the proofs in other PR.

File: content/6_Advanced/Critical.mdx [hunk 1]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
-https://tanujkhattar.wordpress.com/2016/01/11/dominator-tree-of-a-directed-graph/
+### Solution
 
-_ https://codeforces.com/gym/100513/problem/L _
-https://open.kattis.com/problems/skiresort
+### Naive approach
+
+Critical nodes in a DAG are commonly known as **dominators**. Let's define $Dom(u)$ as the set of nodes that dominate node $u$.
+
+The dominator of the starting node is itself. The set of dominators for any other node $u$ is the intersection of the set of dominators for all ancestors
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:47]: starting node?

File: content/6_Advanced/Critical.mdx [hunk 2]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+
+$$
+Dom(u)=
+\begin{cases}
+u,\text{ if u is the starting point}\\
+{u}  \cup (\cap_{p \in ancestor(u)} Dom(p))
+\end{cases}
+$$
+
+The following code uses the above rucerrence.
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:58]:
```suggestion
The following code uses the above recurrence.
```
also add implementation header maybe

File: content/6_Advanced/Critical.mdx [hunk 3]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+	for (auto node : dominators) { cout << node << ' '; }
+
+	return 0;
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+**Time complexity:** $\mathcal{O}(N \cdot M)$.
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:116]: put this above the impl

File: content/6_Advanced/Critical.mdx [hunk 4]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+		title="Visualizing Dominators"
+		url="https://sbaziotis.com/compilers/visualizing-dominators.html"
+		starred
+	>
+    Well-covered article
+    </Resource>
+</Resources>
 
-https://tanujkhattar.wordpress.com/2016/01/11/dominator-tree-of-a-directed-graph/
+### Solution
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:41]: make all these headers one less deep (all the way down to line 170 or so)

File: content/6_Advanced/Critical.mdx [hunk 5]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+
+```cpp
+#include <functional>
+#include <iostream>
+#include <algorithm>
+#include <vector>
+
+using namespace std;
+
+int main() {
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:236]: what if i formatted all this code myself at the cost of 5$ off your bounty

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Critical.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:236]: deal

File: content/6_Advanced/Critical.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:58]: done

File: content/6_Advanced/Critical.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:47]: ok

File: content/6_Advanced/Critical.mdx (hunk 3 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:116]: done

File: content/6_Advanced/Critical.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx]: done

File: content/6_Advanced/Critical.mdx [hunk 6]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+
+$$
+Dom(u)=
+\begin{cases}
+u,\text{ if u is the starting point}\\
+{u}  \cup (\cap_{p \in ancestor(u)} Dom(p))
+\end{cases}
+$$
+
+The following code uses the above rucurrence.
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:58]:
```suggestion
The following code uses the above recurrence.
```

File: content/6_Advanced/Critical.mdx [hunk 7]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
-https://open.kattis.com/problems/skiresort
+## Naive approach
+
+Critical nodes in a DAG are commonly known as **dominators**. Let's define $Dom(u)$ as the set of nodes that dominate node $u$.
+
+The dominator of the starting node $1$ is itself. The set of dominators for any other node $u$ is the intersection of the set of dominators for all ancestors
+$p$ of node $u$.
+
+$$
+Dom(u)=
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:51]: texttt{}? i think standard function formatting is to have all lowercase letters right (optional)

File: content/6_Advanced/Critical.mdx [hunk 8]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+
+This snippet of code combines steps $2$ and $3$:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+// Iterate decreasingly
+for (int i = n - 1; i >= 0; i--) {
+	// Iterat over incoming edges applying (2)
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:198]:
```suggestion
	// Iterate over incoming edges applying (2)
```

File: content/6_Advanced/Critical.mdx [hunk 9]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+	for (int p : rg[i]) { sdom[i] = min(sdom[i], sdom[Find(p, 0)]); }
+	// Insert i into it's bucket after sdom[i]
+	if (i > 0) { bucket[sdom[i]].push_back(i); }
+	// Apply (1) for nodes in bucket
+	for (int node : bucket[i]) {
+		int p = Find(node, 0);
+		if (sdom[p] == sdom[node]) {
+			dom[node] = sdom[node];
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:206]: space chunks of code out a little bit

File: content/6_Advanced/Critical.mdx [hunk 10]
```diff
@@ -10,17 +10,318 @@ frequency: 1
...
+				dfs(son);
+				// Store the parent of son in DFS tree
+				parent[e[son]] = e[node];
+			}
+			// Store the reverse graph
+			rg[e[son]].push_back(e[node]);
+		}
+	};
+	// DSU function
+	function<int(int, int)> Find = [&](int u, int v) {
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:271]: lowercase (same with other var names)

**Review by ryanchou-dev — COMMENTED** [review]
nice graph!

File: content/6_Advanced/Critical.mdx (hunk 6 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:58]: ok

File: content/6_Advanced/Critical.mdx (hunk 8 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:198]: done

File: content/6_Advanced/Critical.mdx (hunk 9 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:206]: ok

File: content/6_Advanced/Critical.mdx (hunk 10 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx]: yep

File: content/6_Advanced/Critical.mdx (hunk 7 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:51]: Idk, \texttt uses monospace and it look kinda weird

**Comment by brebenelmihnea** [comment]
Ok I made the changes.

File: content/6_Advanced/Critical.mdx (hunk 9 above)
Comment (ryanchou-dev) [reply content/6_Advanced/Critical.mdx:204]: not fixed?

File: content/6_Advanced/Critical.mdx [hunk 11]
```diff
@@ -10,17 +10,323 @@ frequency: 1
...
+	for (int node : bucket[i]) {
+		int p = Find(node, 0);
+		if (sdom[p] == sdom[node]) {
+			dom[node] = sdom[node];
+		} else {
+			dom[node] = p;
+		}
+	}
+	// Mark i as seen
+	if (i > 0) { Unite(parent[i], i); }
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:210]:
```suggestion
	if (i > 0) { unite(parent[i], i); }
```

**Review by ryanchou-dev — COMMENTED** [review]
cool!

File: content/6_Advanced/Critical.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Critical.mdx:204]: dw ryan i'm boutta sansify this whole module

File: content/6_Advanced/Critical.mdx [hunk 12]
```diff
@@ -1,26 +1,452 @@
...
+```
+
+</CPPSection>
+</LanguageSection>
 
 ## Cycles
 
 - GP of Wroclaw 2020 H
 - [The Meeting Place Cannot be Changed](https://codeforces.com/problemset/problem/982/F)
 - USACO Camp - Acyclic Graphs
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:450]: hm

Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx]: Done

**Comment by brebenelmihnea** [comment]
The code doesn't get AC without standard C I/O

File: content/6_Advanced/Critical.mdx [hunk 13]
```diff
@@ -1,26 +1,526 @@
...
+#include <functional>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+using namespace std;
+
+const int NMAX = 1e5;
+
+unordered_set<int> in[NMAX], out[NMAX];
```
Comment (SansPapyrus683) [inline content/6_Advanced/Critical.mdx:447]: does this actually tle w/ stock set

Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:447]: Yes it does get tle

**Comment by brebenelmihnea** [comment]
image

File: content/6_Advanced/Critical.mdx (hunk 13 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:447]: Take a look:
https://codeforces.com/contest/982/submission/261065993

File: content/6_Advanced/Critical.mdx [hunk 14]
```diff
@@ -1,26 +1,526 @@
...
+\begin{cases}
+  u\quad\text{ if } u \text{ is the starting point} \\
+  {u}\cup \left(\bigcap_{p \in \texttt{ancestor}(u)} \texttt{dom}(p)\right)\quad\text{otherwise}
+\end{cases}
+$$
+
+### Implementation
+
+The following code uses the above recurrence.
+However, it both is too slow and uses too much memory.
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:61]:
```suggestion
However, it's too slow and uses too much memory.
```

File: content/6_Advanced/Critical.mdx [hunk 15]
```diff
@@ -1,26 +1,526 @@
...
+
+We can compute $\texttt{sdom}(u)$ as the minimum node in the intersection of the following groups:
+1. All the nodes $y$ such that there's an edge from $y$ to $u$ and $e(y) \le e(u)$.
+2. All the values of $\texttt{sdom}(x)$ where $x$ is any node such that there's an edge from $u$ to $x$ and $e(x) > e(u)$.
+   To be more mathematically precise, we can define this group as
+   $$
+   \{\texttt{sdom}(x)\ |\ (u, x) \in E\text{ and }e(x)>e(u)\}
+   $$
+
+The proof of why this works is beyond the scope of this module.
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:188]: is there a link where ppl can learn more?

File: content/6_Advanced/Critical.mdx [hunk 16]
```diff
@@ -1,26 +1,526 @@
...
+const int NMAX = 1e5;
+
+unordered_set<int> in[NMAX], out[NMAX];
+
+int main() {
+	::std::ios::sync_with_stdio(false);
+	::std::cin.tie(0);
+	::std::cout.tie(0);
+	int n, m;
+	scanf("%d %d", &n, &m);
```
Comment (ryanchou-dev) [inline content/6_Advanced/Critical.mdx:454]: cin too slow?

**Review by ryanchou-dev — COMMENTED** [review]
looks great so far!

File: content/6_Advanced/Critical.mdx (hunk 16 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:454]: Yes

File: content/6_Advanced/Critical.mdx (hunk 15 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Critical.mdx:188]: Ok I added a link

**Comment by brebenelmihnea** [comment]
@SansPapyrus683 finally merge ?

**Review by SansPapyrus683 — APPROVED** [review]
1. check my edited explanation to see if it still makes sense
2. fill in the ? i put where the time complexity should be

**Comment by SansPapyrus683** [comment]
https://codeforces.com/contest/982/submission/261698419
also no need for proof of ac

**Pull request merged**